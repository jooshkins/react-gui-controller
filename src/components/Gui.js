import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { controllerStyle } from './JSXStyles/controllerStyles';

class Gui extends Component {
	state = {
		hide: this.props.hide,
		drag: false,
		pos: {
			x: 0,
			y: 0
		}
	};

	handleHide = () => {
		this.setState({
			hide: !this.state.hide
		});
	};

	handleData = (path, val) => {
		let data = this.props.data;
		data[path] = val;
		this.props.onUpdate(data);
	};

	handleDown = event => {
		event.persist();
		let { pos } = this.state;
		this.pos = pos;
		this.setState({ drag: true }, () => {
			this.start = {
				x: event.clientX,
				y: event.clientY
			};
			this.handleMove();
		});
	};

	handleMove = () => {
		window.addEventListener('mousemove', e => {
			if (this.state.drag) {
				let xdiff = -(this.start.x - e.clientX) + this.pos.x;
				let ydiff = -(this.start.y - e.clientY) + this.pos.y;
				if (xdiff <= 0) {
					this.setState({
						pos: {
							x: 0,
							y: ydiff
						}
					});
				} else if (ydiff <= 0) {
					this.setState({
						pos: {
							x: xdiff,
							y: 0
						}
					});
				} else {
					this.setState({
						pos: {
							x: xdiff,
							y: ydiff
						}
					});
				}
			}
		});

		window.addEventListener('mouseup', () => {
			this.setState({
				drag: false
			});
		});
	};

	renderChildren = () => {
		const { children, data } = this.props;
		return React.Children.toArray(children).map((child, i) => {
			return React.cloneElement(child, {
				key: i,
				num: i,
				theme: this.props.theme,
				updateData: this.handleData,
				data: data
			});
		});
	};

	render() {
		const style = {
			transform: `translate(${this.state.pos.x}px, ${this.state.pos.y}px)`
		};
		return (
			<div
				id="controller-body"
				style={style}
				className={
					this.props.theme === 'dark'
						? 'controller-body dark'
						: 'controller-body'
				}
			>
				<div className="drag" onMouseDown={this.handleDown} />
				<div className={this.state.hide ? 'container hide' : 'container'}>
					{this.renderChildren()}
				</div>
				<div
					onClick={this.handleHide}
					className={
						this.props.theme === 'dark'
							? 'control-button control-button-dark'
							: 'control-button'
					}
				>
					{this.state.hide ? 'Open Controls' : 'Close Controls'}
				</div>
				<style jsx>{controllerStyle}</style>
			</div>
		);
	}
}
export default Gui;

Gui.propTypes = {
	children: PropTypes.arrayOf(PropTypes.node),
	data: PropTypes.object,
	theme: PropTypes.oneOf(['light', 'dark']),
	hide: PropTypes.bool,
	onUpdate: PropTypes.func.isRequired
};

Gui.defaultProps = {
	theme: 'light',
	hide: false
};

export { default as GuiString } from './GuiString';
export { default as GuiNumber } from './GuiNumber';
export { default as GuiSelect } from './GuiSelect';
export { default as GuiEasing } from './GuiEasing';
export { default as GuiColor } from './GuiColor';
export { default as GuiBool } from './GuiBool';
export { default as GuiButton } from './GuiButton';
