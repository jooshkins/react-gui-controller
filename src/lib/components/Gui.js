import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { controllerStyle } from './JSXStyles/controllerStyles';

class Gui extends Component {
    handleData = (path, val) => {
        let data = this.props.data;
        data[path] = val;
        this.props.onUpdate(data);
        this.props.onChange(path);
    };

    renderChildren = () => {
        const { children, data } = this.props;
        return React.Children.toArray(children).map((child, i) => {
            return React.cloneElement(child, {
                key: i,
                num: i,
                theme: this.props.theme,
                updateData: this.handleData,
                data: data
            });
        });
    };

    render() {
        return (
            <div
                id="controller-body"
                className={
                    this.props.theme === 'dark'
                        ? 'controller-body dark'
                        : 'controller-body'
                }
            >
                <div>
                    {this.renderChildren()}
                </div>
                <style jsx>{controllerStyle}</style>
            </div>
        );
    }
}

export default Gui;

Gui.propTypes = {
    children: PropTypes.array,
    data: PropTypes.object,
    theme: PropTypes.oneOf(['light', 'dark']),
    hide: PropTypes.bool,
    onUpdate: PropTypes.func.isRequired
};

Gui.defaultProps = {
    theme: 'light',
    hide: false
};
