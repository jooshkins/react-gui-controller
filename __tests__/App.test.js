import React from 'react';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';

configure({
	adapter: new Adapter()
});
import Gui, {
	GuiString,
	GuiColor,
	GuiBool,
	GuiButton,
	GuiNumber,
	GuiEasing,
	GuiSelect
} from '../src/lib/index';

let data = {
	div: 10,
	mirror: true,
	rainbow: false,
	background: '#f9f7f7',
	color: {
		r: 96,
		g: 239,
		b: 184
	}
};
const update = dat => {
	data = dat;
};

const wrapper = shallow(
	<Gui data={data} onUpdate={update}>
		<GuiNumber path="div" label="Divisions" min={2} max={40} step={1} />
	</Gui>
);

describe('<Gui/>', () => {
	it('renders the Gui', () => {
		wrapper;
	});
});

describe('<Gui/> Snapshot', () => {
	it('creates a snapshot', () => {
		wrapper;
		expect(toJson(wrapper)).toMatchSnapshot();
	});
});
