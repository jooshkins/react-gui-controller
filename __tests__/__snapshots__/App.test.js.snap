// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Gui/> Snapshot creates a snapshot 1`] = `
<div
  className="jsx-1631976848 controller-body"
  id="controller-body"
  style={
    Object {
      "transform": "translate(0px, 0px)",
    }
  }
>
  <div
    className="jsx-1631976848 drag"
    onMouseDown={[Function]}
  />
  <div
    className="jsx-1631976848 container"
  >
    <GuiNumber
      data={
        Object {
          "background": "#f9f7f7",
          "color": Object {
            "b": 184,
            "g": 239,
            "r": 96,
          },
          "div": 10,
          "mirror": true,
          "rainbow": false,
        }
      }
      key="0"
      label="Divisions"
      max={40}
      min={2}
      num={0}
      path="div"
      step={1}
      theme="light"
      updateData={[Function]}
    />
  </div>
  <div
    className="jsx-1631976848 control-button"
    onClick={[Function]}
  >
    Close Controls
  </div>
  <JSXStyle
    css=".controller-body.jsx-1631976848{position:fixed;margin:0;padding:0;box-sizing:border-box;width:300px;height:auto;background:rgb(250,250,250);border:1px solid rgb(214,214,214);border-radius:4px;overflow:hidden;z-index:100;}.dark.jsx-1631976848{background:#212121;border:1px solid #313131;}.container.jsx-1631976848{position:relative;margin:0;padding:0;box-sizing:border-box;width:100%;-webkit-transition:0.4s;transition:0.4s;max-height:calc(100vh - 54px);overflow:scroll;}.hide.jsx-1631976848{max-height:0px;}.drag.jsx-1631976848{position:relative;width:100%;height:24px;cursor:all-scroll;}.control-button.jsx-1631976848{position:relative;margin:0;padding:0;box-sizing:border-box;width:100%;height:30px;text-align:center;color:rgb(119,155,255);font-family:sans-serif;font-size:12px;padding-top:8px;font-weight:100;cursor:pointer;border-top:1px solid rgb(214,214,214);}.control-button-dark.jsx-1631976848{border-top:1px solid #313131;}"
    styleId="1631976848"
  />
</div>
`;
